Framework/DataHandling/src/ProcessBankCompressed.cpp

```diff
@@ -0,0 +1,274 @@
+    msg << "Encountered invalid detid=" << detid;
+    throw std::runtime_error(msg.str());
+  }
+
+  const auto det_index = static_cast<size_t>(detid - m_detid_min);
+  m_spectra_accum[period_index][det_index]->createWeightedEvents(raw_events);
+
+  // let go of the unique_ptr to free up memory
+  m_spectra_accum[period_index][det_index].reset();
+}
+
+namespace { // anonymous
+// private class to allow TBB do a more controlled spreading out of event creation
+// since each detid/spectrum/EventList is already independent, these can be done at once
+class EventCreationTask {
+public:
+  EventCreationTask(std::vector<std::unique_ptr<DataHandling::CompressEventAccumulator>> *accumulators,
+                    std::vector<std::vector<Mantid::DataObjects::WeightedEventNoTime> *> *eventlists,
+                    const detid_t detid_min, std::vector<DataObjects::EventSortType> *sorting)
+      : m_accumulators(accumulators), m_eventlists(eventlists), m_sorting(sorting),
+        m_detid_min(static_cast<size_t>(detid_min)) {}
+
+  void operator()(const tbb::blocked_range<size_t> &range) const {
+    for (size_t index = range.begin(); index < range.end(); ++index) {
+      auto &accumulator = m_accumulators->operator[](index);
+      if (accumulator->totalWeight() > 0.) {
+        auto *raw_events = m_eventlists->operator[](index + m_detid_min);
+        // create the events on the correct event list
+        m_accumulators->operator[](index)->createWeightedEvents(raw_events);
+        // drop extra space if the capacity is more than 10% of what is needed
+        if (static_cast<double>(raw_events->capacity()) > 1.1 * static_cast<double>(raw_events->size()))
+          raw_events->shrink_to_fit();
+      }
+      // get the sorting type back
+      m_sorting->operator[](index) = m_accumulators->operator[](index)->getSortType();
+      // let go of the unique_ptr to free up memory
+      m_accumulators->operator[](index).reset();
+    }
+  }
+
+private:
+  std::vector<std::unique_ptr<DataHandling::CompressEventAccumulator>> *m_accumulators;
+  std::vector<std::vector<Mantid::DataObjects::WeightedEventNoTime> *> *m_eventlists;
+  std::vector<DataObjects::EventSortType> *m_sorting;
+  const size_t m_detid_min;
+};
+
+} // namespace
+
+void ProcessBankCompressed::addToEventLists() {
+  Kernel::Timer timer;
+  const auto num_periods = m_loader.m_ws.nPeriods();
+  const auto num_dets = static_cast<size_t>(m_detid_max - m_detid_min + 1);
+  // loop over periods
+  for (size_t period_index = 0; period_index < num_periods; ++period_index) {
+    // create the events and add them to the EventLists
+    EventCreationTask create_task(&(m_spectra_accum[period_index]), &m_loader.weightedNoTimeEventVectors[period_index],
+                                  m_detid_min, &m_sorting);
+    // grainsize selected to balance overhead of creating threads with how much work is done in a thread
+    const size_t grainsize = std::min<size_t>(20, num_dets / 20);
```

Comment:
As a method in `ProcessBankCompressed` or as a general utility sort of thing? I've included the grain-size calculation in a couple of places because the default grain-size varies based on what is being looped over. I'm not sure how universal of a function it would be.